In [2]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)

In [3]:
file_path = r"C:\Users\Jina\Documents\Retail-Sales\Data\Processed.xlsx"

excel = pd.ExcelFile(file_path)

excel.sheet_names

['product_details', 'retail_data1', 'retail_data2']

In [4]:
retail1 = pd.read_excel(file_path,sheet_name="retail_data1")
retail2 = pd.read_excel(file_path, sheet_name="retail_data2")

products = pd.read_excel(file_path,sheet_name="product_details")

Dataset Shapes

In [5]:
print("Retail Data 1:", retail1.shape)
print("Retail Data 2:", retail2.shape)
print("Product Details:", products.shape)

Retail Data 1: (4243, 16)
Retail Data 2: (4251, 16)
Product Details: (10, 4)


In [6]:
retail1.columns.tolist()

['transaction_id',
 'customer_id',
 'customer_name',
 'product_id',
 'price',
 'product_name',
 'category',
 'purchase_location',
 'city',
 'transaction_date',
 'quantity',
 'payment_method',
 'discount',
 'email',
 'phone',
 'payment_status']

In [7]:
retail2.columns.tolist()

['transaction_id',
 'customer_id',
 'customer_name',
 'product_id',
 'price',
 'product_name',
 'category',
 'purchase_location',
 'city',
 'transaction_date',
 'quantity',
 'payment_method',
 'discount',
 'email',
 'phone',
 'payment_status']

In [8]:
products.columns.tolist()

['product_id', 'product_name', 'category', 'price']

In [9]:
retail1.head()

,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status
0,1,642,Troy Mitchell,108,15000.0,Mixer grinder,Home Appliances,offline,Bangalore,2025-12-06 00:00:00,3,UPI,0.10,Troy60@gmail.com,8385276968,successful
1,2,881,Sarah Guerrero,102,70000.0,Phone,ELEC,online,Bangalore,2025-07-21 00:00:00,5,Card,0.35,SarahGuerrero@yahoo.com,7147248911,successful
2,3,505,Samantha Hull,109,65000.0,Refrigerator,Home Appliances,offline,Chennai,2025-07-11 00:00:00,2,UPI,0.05,SamanthaHull@outlook.com,7415321565,successful
3,4,794,Gerald Cooper,106,80000.0,sofa,FURN,offline,Chennai,2025-10-06 00:00:00,2,UPI,0.05,GeraldCooper@gmail.com,9739354201,successful
4,5,864,Cameron Black,108,15000.0,Mixer grinder,home appliances,offline,Chennai,2025-12-12 00:00:00,1,Cash,0.05,CameronBlack@yahoo.com,9938169477,successful


In [10]:
products.head()

,product_id,product_name,category,price
0,101,Laptop,Electronics,250000
1,102,Phone,Electronics,70000
2,103,Shirt,Clothing,1300
3,104,Shoes,Clothing,8000
4,105,TV,Electronics,45000


## Missing Value Analysis

In [11]:
retail1.isnull().sum().sort_values(ascending=False)

price                404
transaction_id         0
customer_name          0
customer_id            0
product_id             0
product_name           0
category               0
purchase_location      0
city                   0
transaction_date       0
quantity               0
payment_method         0
discount               0
email                  0
phone                  0
payment_status         0
dtype: int64

In [12]:
retail2.isnull().sum().sort_values(ascending=False)

price                405
transaction_id         0
customer_name          0
customer_id            0
product_id             0
product_name           0
category               0
purchase_location      0
city                   0
transaction_date       0
quantity               0
payment_method         0
discount               0
email                  0
phone                  0
payment_status         0
dtype: int64

In [13]:
products.isnull().sum().sort_values(ascending=False)

product_id      0
product_name    0
category        0
price           0
dtype: int64

## Duplicate Analysis

In [14]:
retail1.duplicated().sum()

np.int64(0)

In [15]:
retail2.duplicated().sum()

np.int64(0)

In [16]:
retail1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4243 entries, 0 to 4242
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   transaction_id     4243 non-null   int64  
 1   customer_id        4243 non-null   int64  
 2   customer_name      4243 non-null   object 
 3   product_id         4243 non-null   int64  
 4   price              3839 non-null   float64
 5   product_name       4243 non-null   object 
 6   category           4243 non-null   object 
 7   purchase_location  4243 non-null   object 
 8   city               4243 non-null   object 
 9   transaction_date   4243 non-null   object 
 10  quantity           4243 non-null   int64  
 11  payment_method     4243 non-null   object 
 12  discount           4243 non-null   float64
 13  email              4243 non-null   object 
 14  phone              4243 non-null   int64  
 15  payment_status     4243 non-null   object 
dtypes: float64(2), int64(5),

In [17]:
retail2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4251 entries, 0 to 4250
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   transaction_id     4251 non-null   int64  
 1   customer_id        4251 non-null   int64  
 2   customer_name      4251 non-null   object 
 3   product_id         4251 non-null   int64  
 4   price              3846 non-null   float64
 5   product_name       4251 non-null   object 
 6   category           4251 non-null   object 
 7   purchase_location  4251 non-null   object 
 8   city               4251 non-null   object 
 9   transaction_date   4251 non-null   object 
 10  quantity           4251 non-null   int64  
 11  payment_method     4251 non-null   object 
 12  discount           4251 non-null   float64
 13  email              4251 non-null   object 
 14  phone              4251 non-null   int64  
 15  payment_status     4251 non-null   object 
dtypes: float64(2), int64(5),

In [18]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   product_id    10 non-null     int64 
 1   product_name  10 non-null     object
 2   category      10 non-null     object
 3   price         10 non-null     int64 
dtypes: int64(2), object(2)
memory usage: 452.0+ bytes


In [19]:
retail1.describe(include='all')

,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status
count,4243.000000,4243.000000,4243,4243.000000,3839.000000,4243,4243,4243,4243,4243,4243.000000,4243,4243.000000,4243,4.243000e+03,4243
unique,NaN,NaN,973,NaN,NaN,30,12,2,5,643,NaN,4,NaN,977,NaN,2
top,NaN,NaN,Amanda Castillo,NaN,NaN,Phone,ELEC,offline,Chennai,2026-02-02 00:00:00,NaN,UPI,NaN,AmandaCastillo@yahoo.com,NaN,successful
freq,NaN,NaN,15,NaN,NaN,158,470,2155,912,19,NaN,1079,NaN,15,NaN,4000
mean,1998.654018,499.950271,NaN,105.470893,64982.417296,NaN,NaN,NaN,NaN,NaN,2.936130,NaN,0.221400,NaN,7.969172e+09,NaN
std,1155.006318,292.839452,NaN,2.902760,69565.038424,NaN,NaN,NaN,NaN,NaN,1.474096,NaN,0.114003,NaN,1.157202e+09,NaN
min,1.000000,1.000000,NaN,101.000000,1300.000000,NaN,NaN,NaN,NaN,NaN,-2.000000,NaN,0.050000,NaN,6.003830e+09,NaN
25%,995.500000,244.000000,NaN,103.000000,15000.000000,NaN,NaN,NaN,NaN,NaN,2.000000,NaN,0.100000,NaN,6.962432e+09,NaN
50%,1996.000000,502.000000,NaN,105.000000,45000.000000,NaN,NaN,NaN,NaN,NaN,3.000000,NaN,0.200000,NaN,7.954633e+09,NaN
75%,3000.500000,757.000000,NaN,108.000000,70000.000000,NaN,NaN,NaN,NaN,NaN,4.000000,NaN,0.300000,NaN,8.961772e+09,NaN


In [20]:
retail2.describe(include='all')

,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status
count,4251.000000,4251.000000,4251,4251.000000,3846.000000,4251,4251,4251,4251,4251,4251.000000,4251,4251.000000,4251,4.251000e+03,4251
unique,NaN,NaN,976,NaN,NaN,30,12,2,5,645,NaN,4,NaN,982,NaN,2
top,NaN,NaN,Jason Taylor,NaN,NaN,LAPTOP,ELEC,offline,Delhi,2025-06-12 00:00:00,NaN,Card,NaN,AmandaHuff@gmail.com,NaN,successful
freq,NaN,NaN,14,NaN,NaN,165,442,2159,929,19,NaN,1096,NaN,12,NaN,4000
mean,5999.695601,1495.326041,NaN,105.481063,63582.059282,NaN,NaN,NaN,NaN,NaN,2.942837,NaN,0.227970,NaN,7.928919e+09,NaN
std,1153.577974,285.525374,NaN,2.867802,68806.010795,NaN,NaN,NaN,NaN,NaN,1.458211,NaN,0.115483,NaN,1.106933e+09,NaN
min,4001.000000,1001.000000,NaN,101.000000,1300.000000,NaN,NaN,NaN,NaN,NaN,-2.000000,NaN,0.050000,NaN,6.000831e+09,NaN
25%,5001.500000,1252.500000,NaN,103.000000,15000.000000,NaN,NaN,NaN,NaN,NaN,2.000000,NaN,0.150000,NaN,6.991124e+09,NaN
50%,6003.000000,1488.000000,NaN,106.000000,50000.000000,NaN,NaN,NaN,NaN,NaN,3.000000,NaN,0.250000,NaN,7.899737e+09,NaN
75%,6997.500000,1740.000000,NaN,108.000000,70000.000000,NaN,NaN,NaN,NaN,NaN,4.000000,NaN,0.350000,NaN,8.839547e+09,NaN


In [21]:
profile = pd.DataFrame({'Column': retail1.columns,'DataType': retail1.dtypes.values,'MissingValues': retail1.isnull().sum().values})
profile

,Column,DataType,MissingValues
0,transaction_id,int64,0
1,customer_id,int64,0
2,customer_name,object,0
3,product_id,int64,0
4,price,float64,404
5,product_name,object,0
6,category,object,0
7,purchase_location,object,0
8,city,object,0
9,transaction_date,object,0


## Data Quality Issue 1: Missing Price Values

Observation:
- retail_data1 contains 404 missing values in the price column.
- retail_data2 contains 405 missing values in the price column.

Business Impact:
- Revenue calculations will be inaccurate.
- Category and product performance metrics will be affected.

Resolution:
- Missing prices will be populated using the reference product_details table based on product_id.

In [22]:
products.head()

,product_id,product_name,category,price
0,101,Laptop,Electronics,250000
1,102,Phone,Electronics,70000
2,103,Shirt,Clothing,1300
3,104,Shoes,Clothing,8000
4,105,TV,Electronics,45000


In [23]:
products['product_id'].nunique()

10

In [24]:
products.shape

(10, 4)

In [25]:
retail1.duplicated().sum()

np.int64(0)

In [26]:
retail2.duplicated().sum()

np.int64(0)

In [27]:
retail1['transaction_id'].duplicated().sum()

np.int64(243)

In [28]:
retail2['transaction_id'].duplicated().sum()

np.int64(251)

In [29]:
(retail1['quantity'] <= 0).sum()

np.int64(50)

In [30]:
(retail2['quantity'] <= 0).sum()

np.int64(43)

In [31]:
retail1[retail1['quantity'] <= 0]

,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status
33,33,50,Stacey Henderson,104,8000.0,SHOES,CLOTH,online,Mumbai,2025-06-14 00:00:00,-1,Card,0.40,Stacey33@gmail.com,6033182385,successful
75,74,907,Jeffrey Mccall,102,70000.0,PHONE,ELEC,online,Hyderabad,2025-05-26 00:00:00,0,Card,0.40,Jeffrey63@gmail.com,6709574650,successful
77,76,335,Jacob Evans,103,1300.0,shirt,CLOTH,offline,Mumbai,2026-04-11 00:00:00,-2,NetBanking,0.35,JacobEvans@yahoo.com,6483959310,successful
78,77,780,Elizabeth Miller,110,NaN,MICROWAVE,Home Appliances,offline,Delhi,2025-09-14 00:00:00,0,Card,0.30,ElizabethMiller@yahoo.com,8279569798,successful
100,97,723,Christopher Walter,106,80000.0,SOFA,furniture,offline,Hyderabad,2026-02-27 00:00:00,0,NetBanking,0.15,ChristopherWalter@gmail.com,9036645707,successful
139,133,633,Julie Jackson,103,1300.0,shirt,Clothing,offline,Delhi,2025-07-13 00:00:00,-1,UPI,0.20,Julie50@gmail.com,7168243701,successful
356,339,758,Nichole Jones,108,15000.0,MIXER GRINDER,home appliances,offline,Chennai,2025-10-07 00:00:00,0,UPI,0.30,NicholeJones@yahoo.com,6141266772,successful
661,624,925,Steve Ayala,101,250000.0,laptop,Electronics,offline,Hyderabad,07-14-2025,0,NetBanking,0.40,SteveAyala@gmail.com,9576287018,successful
714,674,826,Ryan Wade,102,70000.0,Phone,Electronics,online,Mumbai,2026-01-07 00:00:00,-1,Cash,0.40,RyanWade@outlook.com,7368883000,successful
893,841,725,Jody Martinez,105,45000.0,TV,Electronics,online,Bangalore,04-19-2025,0,NetBanking,0.40,JodyMartinez@gmail.com,7989096889,successful


In [32]:
sorted(retail1['category'].dropna().unique())

['CLOTH',
 'Clothing',
 'ELEC',
 'Electronics',
 'FURN',
 'Furniture',
 'HOME',
 'Home Appliances',
 'clothing',
 'electronics',
 'furniture',
 'home appliances']

In [33]:
retail1['payment_status'].value_counts(dropna=False)

payment_status
successful    4000
failed         243
Name: count, dtype: int64

In [34]:
sorted(retail2['category'].dropna().unique())

['CLOTH',
 'Clothing',
 'ELEC',
 'Electronics',
 'FURN',
 'Furniture',
 'HOME',
 'Home Appliances',
 'clothing',
 'electronics',
 'furniture',
 'home appliances']

In [35]:
retail2['payment_status'].value_counts(dropna=False)

payment_status
successful    4000
failed         251
Name: count, dtype: int64

In [36]:
sorted(retail1['product_name'].dropna().unique())

['DINING TABLE',
 'Dining table',
 'LAPTOP',
 'Laptop',
 'MICROWAVE',
 'MIXER GRINDER',
 'Microwave',
 'Mixer grinder',
 'PHONE',
 'Phone',
 'REFRIGERATOR',
 'Refrigerator',
 'SHIRT',
 'SHOES',
 'SOFA',
 'Shirt',
 'Shoes',
 'Sofa',
 'TV',
 'Tv',
 'dining table',
 'laptop',
 'microwave',
 'mixer grinder',
 'phone',
 'refrigerator',
 'shirt',
 'shoes',
 'sofa',
 'tv']

In [37]:
sorted(retail2['product_name'].dropna().unique())

['DINING TABLE',
 'Dining table',
 'LAPTOP',
 'Laptop',
 'MICROWAVE',
 'MIXER GRINDER',
 'Microwave',
 'Mixer grinder',
 'PHONE',
 'Phone',
 'REFRIGERATOR',
 'Refrigerator',
 'SHIRT',
 'SHOES',
 'SOFA',
 'Shirt',
 'Shoes',
 'Sofa',
 'TV',
 'Tv',
 'dining table',
 'laptop',
 'microwave',
 'mixer grinder',
 'phone',
 'refrigerator',
 'shirt',
 'shoes',
 'sofa',
 'tv']

In [38]:
retail1['transaction_date'].sample(20)

1807             05-22-2025
3124    2025-11-17 00:00:00
1704             03-25-2026
4146    2026-01-06 00:00:00
4192    2025-10-15 00:00:00
4077    2025-08-27 00:00:00
403     2026-02-21 00:00:00
641              12-30-2025
1373    2025-04-16 00:00:00
442     2025-11-10 00:00:00
3173    2025-08-12 00:00:00
1410    2026-04-02 00:00:00
3321             03-22-2026
2409    2025-07-05 00:00:00
782              12-29-2025
294     2025-10-23 00:00:00
480     2025-03-07 00:00:00
3618    2025-09-09 00:00:00
3167    2025-06-25 00:00:00
829     2025-04-27 00:00:00
Name: transaction_date, dtype: object

In [39]:
retail2['transaction_date'].sample(20)

1851             04-21-2025
1250    2025-11-02 00:00:00
4184    2026-03-17 00:00:00
3719    2026-01-04 00:00:00
2960             03-28-2026
2767    2025-12-19 00:00:00
2017             08-21-2025
2958    2025-01-10 00:00:00
21      2025-05-07 00:00:00
2438    2026-01-05 00:00:00
1353             10-23-2025
1772    2025-04-08 00:00:00
851              02-14-2026
1163    2026-01-16 00:00:00
2519    2026-01-11 00:00:00
3431             08-28-2025
3713    2025-07-27 00:00:00
2255    2026-03-03 00:00:00
859              12-15-2025
989     2026-02-06 00:00:00
Name: transaction_date, dtype: object

In [40]:
retail1[
    retail1['transaction_id'].duplicated(keep=False)
].sort_values('transaction_id')

,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status
5,6,181,Maria Robinson,101,250000.0,laptop,Electronics,online,Delhi,2025-07-25 00:00:00,5,NetBanking,0.35,MariaRobinson@gmail.com,8794772717,successful
6,6,181,Maria Robinson,101,250000.0,laptop,Electronics,online,Delhi,2025-07-25 00:00:00,5,NetBanking,0.35,MariaRobinson@gmail.com,8794772717,failed
57,57,72,Michael Wood,105,45000.0,tv,ELEC,online,Chennai,12-31-2025,5,Cash,0.20,Michael77@gmail.com,8385570077,successful
58,57,72,Michael Wood,105,45000.0,tv,ELEC,online,Chennai,12-31-2025,5,Cash,0.20,Michael77@gmail.com,8385570077,failed
83,82,578,Kristi Wilson,101,250000.0,Laptop,Electronics,online,Chennai,2025-06-18 00:00:00,4,NetBanking,0.40,KristiWilson@yahoo.com,6043009323,successful
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4182,3943,774,Antonio Evans,106,80000.0,Sofa,FURN,offline,Hyderabad,09-28-2025,4,Cash,0.05,Antonio94@gmail.com,7199037042,successful
4187,3946,504,Victor Hendricks,102,70000.0,Phone,ELEC,offline,Mumbai,2026-01-21 00:00:00,4,Card,0.30,VictorHendricks@gmail.com,6589400307,failed
4186,3946,504,Victor Hendricks,102,70000.0,Phone,ELEC,offline,Mumbai,2026-01-21 00:00:00,4,Card,0.30,VictorHendricks@gmail.com,6589400307,successful
4230,3988,933,Zachary Khan,104,8000.0,shoes,CLOTH,online,Bangalore,2025-06-28 00:00:00,4,Cash,0.05,Zachary6@gmail.com,9316003581,failed


In [41]:
retail1[retail1['quantity'] <= 0]

,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status
33,33,50,Stacey Henderson,104,8000.0,SHOES,CLOTH,online,Mumbai,2025-06-14 00:00:00,-1,Card,0.40,Stacey33@gmail.com,6033182385,successful
75,74,907,Jeffrey Mccall,102,70000.0,PHONE,ELEC,online,Hyderabad,2025-05-26 00:00:00,0,Card,0.40,Jeffrey63@gmail.com,6709574650,successful
77,76,335,Jacob Evans,103,1300.0,shirt,CLOTH,offline,Mumbai,2026-04-11 00:00:00,-2,NetBanking,0.35,JacobEvans@yahoo.com,6483959310,successful
78,77,780,Elizabeth Miller,110,NaN,MICROWAVE,Home Appliances,offline,Delhi,2025-09-14 00:00:00,0,Card,0.30,ElizabethMiller@yahoo.com,8279569798,successful
100,97,723,Christopher Walter,106,80000.0,SOFA,furniture,offline,Hyderabad,2026-02-27 00:00:00,0,NetBanking,0.15,ChristopherWalter@gmail.com,9036645707,successful
139,133,633,Julie Jackson,103,1300.0,shirt,Clothing,offline,Delhi,2025-07-13 00:00:00,-1,UPI,0.20,Julie50@gmail.com,7168243701,successful
356,339,758,Nichole Jones,108,15000.0,MIXER GRINDER,home appliances,offline,Chennai,2025-10-07 00:00:00,0,UPI,0.30,NicholeJones@yahoo.com,6141266772,successful
661,624,925,Steve Ayala,101,250000.0,laptop,Electronics,offline,Hyderabad,07-14-2025,0,NetBanking,0.40,SteveAyala@gmail.com,9576287018,successful
714,674,826,Ryan Wade,102,70000.0,Phone,Electronics,online,Mumbai,2026-01-07 00:00:00,-1,Cash,0.40,RyanWade@outlook.com,7368883000,successful
893,841,725,Jody Martinez,105,45000.0,TV,Electronics,online,Bangalore,04-19-2025,0,NetBanking,0.40,JodyMartinez@gmail.com,7989096889,successful


### Data Quality Issue 2: Invalid Quantities

Observation:
Non-positive quantity values (0, -1, -2) were identified in the transaction datasets.

Business Impact:
These records would distort revenue calculations, product performance metrics, and sales reporting.

Resolution:
Records with quantity <= 0 will be removed from the dataset during the cleaning phase.

### Data Quality Issue 3: Inconsistent Category Values

Observation:
The category column contains multiple representations of the same category such as:
- CLOTH, Clothing, clothing
- ELEC, Electronics, electronics
- FURN, Furniture, furniture
- HOME, Home Appliances, home appliances

Business Impact:
Revenue and sales metrics would be fragmented across multiple category labels, leading to inaccurate reporting.

Resolution:
Standardize category values using a mapping dictionary and convert them to a common format.

### Data Quality Issue 4: Inconsistent Product Names

Observation:
The same products appear in multiple formats.

Examples:
- LAPTOP, Laptop, laptop
- PHONE, Phone, phone
- DINING TABLE, Dining table, dining table

Business Impact:
Product-level sales and revenue analysis would be fragmented across multiple labels.

Resolution:
Standardize product names using the product dimension table and convert values to a consistent format.